### ISSUE TITLE
Configure Web3APIs Using Env Variables

### ISSUE THREAD

**Issue by dOrgJelli** [Web3-API/prototype#140]
After talking with Luis from Pocket Network, it's become clear that we'll need a way to set configuration values for Web3APIs.

## Current Problem
There is currently one way to configure Web3APIs, and that's through method arguments. For example, think of an IPFS HTTP Client Web3API, you'd have to tell it what endpoint to talk to via a method argument:
```graphql
type Query {
  cat(
    hash: String!
    node: String!
  ): Bytes!
}
```

While this is simple and it works, this becomes problematic when you think about composability. If the IPFS Web3API is being used within another Web3API "Foo", you'd have to pass in the `node: String!` into each of Foo's methods, so it can pass it through to IPFS:
```graphql
#import { Query } into IPFS from "w3://ens/ipfs-http-client.web3api.eth"

type Query {
  doFoo(
    fooArg: [UInt32!]!
    ipfsNode: String!
  )
}
```

If you think about scaling this farther, it's clearly an anti-pattern, leading to N number of arguments needed to configure N number of dependencies.

NOTE: Web3APIs do not have persistent state (heap), meaning that the node URL cannot be stored in a global and used for subsequent method invocations.

## Proposed Solution
Enable the configuration of Web3APIs through env configuration variables. In the example above, this would allow the user to define an IPFS node's endpoint once.

## Developer Experience
*Still Very WIP...*

1. Web3API developer defines ENV variables using the schema.graphql
```graphql
type Env {
  requiredSetting: String!
  optionalSetting: UInt
}
```

2. Web3API developer uses ENV variables in WASM code
```typescript
import { Env } from "./w3";

Env.requiredSetting
```

3. Users configure the Web3API through the Web3API Client
```typescript
new Web3ApiClient({
  constants: [
    {
      uri: "w3://ens/some.api.eth",
      env: {
        requiredString: "Something Important",
        optionalString: nullableValue
      }
    }
  ]
});
```

Additional usability thoughts here: https://github.com/Web3-API/monorepo/issues/140#issuecomment-782519627

## Worries
Does this create "configuration hell"?

**Comment by mpetrunic** [comment]
Can't you just configure ipfs node url  when initializing plugin/package?

**Comment by dOrgJelli** [comment]
Yes this is possible with plugins, but we haven't designed a way to do this for WASM based Web3APIs. That's the point of this issue.

**Comment by mpetrunic** [comment]
Are WASM based Web3APIs allowed to have state? I mean package could export some init method for configuration so we leave this as implementation detail of web3api developers?

**Comment by dOrgJelli** [comment]
No all WASM Web3APIs are stateless @mpetrun5. The heap is created and destroyed for each individual invocation.

**Comment by dOrgJelli** [comment]
After talking about this feature in the Core Dev Sync today, it looks like there are no objections and all voiced opinions seem to think this is a necessary next step.

**Comment by dOrgJelli** [comment]
Additional detail:
- These env variables should be configurable at both initialization-time, and query-time. This will be useful for sub-query configuration. For example, if the ENS Web3API is querying the Rinkeby network, but the ethereum's default provider has been set to mainnet. When ENS queries Ethereum, it can configure its provider using some default used for rinkeby.

**Comment by dOrgJelli** [comment]
An example of this anti-pattern described above (configuration through method arguments) can be found in the SimpleStorage demo, which uses the Ethereum Web3API as a dependency.

Starting from the root of the problem, the Ethereum plugin's schema allows callers to configure a `Connection`: https://github.com/Web3-API/monorepo/blob/ac7c58ac311f4c1b9f1e632b2dd0957184bdf5bc/packages/js/plugins/ethereum/schema.graphql#L26-L29

This connection interface can be used to specify which network the Ethereum plugin should be using (mainnet, rinkeby, testnet, etc), as well as overriding which `node` RPC endpoint it should be making requests to. The connection is configured using an optional method argument on the Ethereum module's methods, for example see `callView`: https://github.com/Web3-API/monorepo/blob/ac7c58ac311f4c1b9f1e632b2dd0957184bdf5bc/packages/js/plugins/ethereum/schema.graphql#L1-L8

Now to understand where this becomes a problem, we must look at a user of the Ethereum Web3API.

In SimpleStorage's Web3API (found [here](https://github.com/Web3-API/demos/tree/main/simple-storage/web3api)), it uses Ethereum under the hood. In order to allow a user of SimpleStorage to configure the Ethereum dependency, an additional argument must be added [like so](https://github.com/Web3-API/demos/blob/8b0b3c99b00bc456f25a279889c577baaf876df9/simple-storage/web3api/src/query/schema.graphql#L1-L8):
```graphql
#import { Query, Connection } into Ethereum from "w3://ens/ethereum.web3api.eth"

type Query {
  getData(
    address: String!
    connection: Ethereum_Connection
  ): Int!
}
```

Users of SimpleStorage can specify Ethereum's configuration like so:
```typescript
client.query({
  uri: "ens/api.simplestorage.eth",
  query: `{
    getData(
      address: "0x..."
      connection: {
        networkNameOrChainId: "rinkeby"
      }
    )
  }`
})
```

**Comment by pwvpwvpwv** [comment]
After discussions with @dOrgJelli we came up the following implementation that we'd like to get some feedback/discussions goinh on.
## Requirements
* Schema 
	* Schema will be defined in either the `src/mutations/schema.graphql` file or the `src/queries/schema.graphql` file for the API application 
	* Schema for Environmental Variables are defined leveraging GraphQL Schema definition using a reserved type named `Env`.
		* Allows for definition of a required field using GraphQL Schema required - `!` - notation. 
		* Allows for definition of public variables exposed to Web3API clients via custom GraphQL directive `@public()`.
		* Allows default values to be set on optional fields using GraphQL argument syntax.
		* All values sent will be sanitized by verifying the values match the schema such as making sure all required fields are set, and types match the schema automatically.
	* A `Client_Env` reserved type can be used to define what the Environmental Variables shape is from the client perspective - and in this case the reserved type `Env` can include additional values not required by the ENV (defaults), or provide a way to modify the initial environmental variables payload.
	* A reserved `Env` query will be defined to retrieve the sanitized environmental variable values and allow for a custom hook to be implemented to further validate/change values.
* Definition of Environmental Variables
	* Enviromental variables are specified in a constants argument when initializing the Web3API client.
	* Environmental variables can be defined as either `env` or `mut_env` to specify whether the variables can only be used in queries or mutations - this is a self governing system and all values will be defined on the `Env` reserved type and split up internally during the build process into `Env` and `Mut_Env`. 
* Accessing Environmental Variables
	* Imports from app/client variables can be imported using the `Env` variable provided by `w3` module. 
	* Environmental variables exposed via the `@public()` directive will be importable by other W3 APIs.
	* As environmental variables are cached on the client side, values should be able to be pulled in individually as needed via a function call to allow for deserialization of necessary values - not the entire cached environmental variables blob.

## Technical Implementation Details
### Schema & Sanitization
Schema definitions will leverage existing GraphQL schema file that identifies endpoints for Web3 applications. The idea here is allow simple way to define these variables, and a more custom way to allow developers to implement their own solution if the exposed variables end up needing to be modified or change in some way outside of GraphQL schema capabilities. Below is an example of a simple schema containing a required setting, an optional setting:
```graphql
type Env {
  requiredSetting: String!
  optionalSetting: String
}
```
Implementing the reserved `Client_Env` will allow for changing of the values that end up being exposed in the Web3 app. Below is an example of the schema implementing the `Env,` and also marking a value as public allowing it to be exposed to other Web3 APIs:
```graphql
type Query { 
  Env(env: ClientEnv!): Env!  
}  

# pre sanitization 
type ClientEnv {  
  requiredSetting: String!
  optionalSetting: String
}  

# post sanitization 
type Env {  
  sanitizedRequiredSetting: String!  
  sanitizedOptionalSetting: String! @public()
  alwaysSetViaSanitization: UInt! 
}
```
The sanitization of the client values can now include a `Env` query to be written which will modify the values and shape of the Environmental Variables as specified. This allows for any custom logic to be built when creating these environments that the developer chooses. Below is a rough idea of how a custom sanitize hook might look in typescript:
```typescript
import { sanitizeObj, // helper method - returns {error: bool, value: message/data}  
  failSanitization,  
  succeedSanitization  
} from './w3';  
  
// Simple validator requiring https (not http:)
sanitizeRequired(k, str) => {  
  return str.slice(0, 5) !== 'https'  
    ? sanitizeObj(true, 'HTTPS required')  
    : sanitizeObj(false, endText)  
}  
  
sanitizeEnv(env) => {  // env will be the initial ENV  
  const errors = [];  
  const sanitizedEnv = {};  
  
  // Sanitization via validation  
  const requiredSetting = sanitizeRequired('requiredSetting', env.requiredSetting)  
  requiredSetting.error  
	? errors.push(requiredSetting)  
    : sanitizedEnv.requiredSetting = requiredSetting;  
  
  // No sanitization or change to this setting  
  sanitizedEnv.optionalSetting = env.optionalSetting;  
  
  // Additional argument added to SanitizedEnv  
  sanitizedEnv.alwaysSetViaSanitization = 3;  
  
  // We call the fail/succeed functions with payload when done  
  errors.length ? failSanitization(errors) : succeedSanitization(sanitizedEnv)'  
}
```
 
### Setting Environmental Values
The values for the Environmental Variables are able to be configured by the client making the request. This is done during the instantiation of the Web3 client. This call could look like this:
```typescript
new Web3ApiClient({
	constants: [
		{
			uri: 'ens/someaddress.eth',
			env: {
				provider: 'localhost:8545'
			}
		}
	]
})
``` 
Also a custom `env.json` or `env.ts` file could be created to remove the need to hard code this data into the codebase, allowing a more configuration approach. This would simply be loaded automatically into the `env` property. The typescript version would allow maximum flexibility and allow the developer writing the client to code to export the variables based on any number of system configurations such as system environmental variables etc.

This APIs GraphQL Schema might look something like this:
```graphql
...
type Query { 
  sanitizeEnv(env: ClientEnv!): Env!  
}  

# pre sanitization 
type ClientEnv {  
  provider: String!
}  

# post sanitization 
type Env {  
  provider: String! @public()
  api_version: String! @public()
  some_secret_key: String!
}
...
```
From here the client can also import and use these variables as needed via the normal import methods on an exposed w3 object.

We also have the ability to define mutation specific environmental variables, which is up to the developer on if they find the need to do this. This allows for separation of concerns and is a individual self maintained rule. Below is an example of this:
```typescript
new Web3ApiClient({
	constants: [
		{
			uri: 'ens/someaddress.eth',
			env: {
				provider: 'localhost:8545'
			},
			mut_env: {
				secretKey: 'abc123'
			}
		}
	]
})
``` 
From a developers standpoint on the API side, the schema definition should not require any changes than having both of these values specified in the `Env`. Internally though, these will be split up and effective have two reserved types based on the developer definition `Env` and `Mut_Env`.

Given the above example of setting the variables, below is how this would look to the API developer:
```graphql
...
type Env {
	provider: String! @public()
	secretKey: String!
}
...
```
During the build process these would then be separated in a way that provider belongs to `Env` and secretKey is a property of `Mut_Env`. After the build this get's broken up by how the value is set in either `env` or `mut_env` by the client:
```graphql
...
type Env {
	provider: String! @public()
}

type Mut_Env{
	secretKey: String!
}
...
```
### Accessing Environment Variables from API
On the API end this already sanitized environment blob would then be sent when the WASM module is called again, and would be accessible in the API. Here we are able to create our own, or import environments from other apps. Below is an example of this:
```graphql
#import { Env, Mut_Env } into Ethereum from 'ens/ethereum.web3api.eth'
"""rest of schema file..."""
```
And then importing the actual variables for use during runtime in the WASM module. The variables should are accessible via getter methods, as it will get the variable from the clients sanitized cache environment on-demand and not have to load the entire environment for single values. Below is an example of this:
```typescript
import { Ethereum_Env, Mut_Ethereum_Env } from './w3';

function customizeProvider() {
    let providerCustomized = "";
	const provider = Ethereum_Env.provider(); // under the hood makes call to client __w3_envVar('provider')
	// Do any custom logic
	return providerCustomized;
}

function makeCallWithSecret(){
	const secret = Mut_Ethereum_Env.secretKey();
	// Make call to some service using secret etc.
	return response;
}
```

**Comment by dOrgJelli** [comment]
Just talked with @mpetrun5 who'll be working on the implementation. Here's what we came up with as notes (a few design choices still left to be made, still rough):

# Steps
## Fully-Built Schema (schema-parser)
1. Support the new schema features in parser
2. Add lots of validation to the parser to make sure Env types are setup correctly
## User-Defined Schema (schema-composer)
3. Support the "magic seperation" of Env types, defined in both Query and Mutation
4. Write tests to make sure users can import the Env (supported: MutationEnv, QueryEnv)
## Web3ApiClient
5. Constructor config
6. Modify invoke logic, when querying for first time, sanitize?
7. Modify invoke logic, when querying, load env
## Wasm/AS
8. Define the WASM standard interface for things like: (_w3_env_sanitize, _w3_env_load, __w3_import_env)
## Web3ApiClient
9. Add support for __w3_import_env
## Generated Code (schema-bind)
10. Define what we want the generated code to look like
11. Create mustache templates to generate the desired code

# Schema Defining UX
`mutation.graphql`
```graphql
type Env {
  prop: String!
}

type ClientEnv {
  foo: String!
}

# "specialize" this type for the mutation module specifically,
# just add Mutation to the front of it.
type MutationEnv {

}

type MutationClientEnv {

}
```

```graphql
#import { MutationEnv } from "above..."


```


# Schema
## "magic seperation"
`./mutation.graphql`
```graphql
type Env {
  foo: String!
}
```

`./query.graphql`
```graphql
type Env {
  foo: String!
  bar: UInt8!
}
```

`BUILD ->`
```graphql
type MutationEnv {
  foo: String!
}

type QueryEnv {
  bar: UInt8!
}
```

## "manual extension"
`./mutation.graphql`
```graphql
#import { Env } from "..."
```

`./query.graphql`
```graphql
#import { Env } from "..."

type QueryEnv {
  bar: UInt8!
}
```

`BUILD ->`
```graphql
type MutationEnv {
  foo: String!
}

type QueryEnv {
  foo: String!
  bar: UInt8!
}
```



# Client Env Runtime

## 1. User configures the client
```typescript
new Web3ApiClient({
  constants: [
    {
      uri: 'ens/someaddress.eth',
      env: {
        provider: 'localhost:8545'
      },
      mutation_env: {
        secretKey: 'abc123'
      }
    }
  ]
})

new Web3ApiClient({
  env: {
    'ens/someaddress.eth': {
      provider: 'localhost:8545'
      mutation_env: {
        secretKey: 'abc123'
      }
    }
  ]
})
```

## 2. User queries the API
```typescript
client.query({ uri: "ens/someaddress.eth" });
```

## 3. Upon querying the API for the first time, the env is sanitized and cached
`inside of Web3ApiClient...`
```typescript
// if (!finalBlob) {
  // Call into WASM to see if it requires env sanitization `if (module[_w3_env_sanitize] !== undefined)`

  // If required, serialize the env object we have for the module to msgpack,
  // and send it to the web3api (_w3_env_sanitize(blob) -> finalBlob)

  // Cache the finalBlob in the Web3ApiClient
//}

// _w3_env_load(finalBlob)
```

# Public Properties

type Env {
  prop: String! @public
  foo: UInt8!
}


// inside of importer
#import { Env } into Ethereum from "..."


// BUILD (schema-compose) ->

type Ethereum_Env {
  prop: String! @public
}



// DEPENDENCY.WASM

// generated, called by the client

```typescript
import { Env } from "./Env"

_w3_import_env(): PublicEnv {
  return {
    prop: Env.prop
  }
}
```

// USER.WASM

```typescript
import { Ethereum_Env } from "..."

Ethereum_Env.prop;

// __w3_import_env("ethereum.eth") -> client -> DEPDENDENCY.wasm
```

?
```typescript
interface IEthereum_Env {
  ...
}

const instance: IEthereum_Env = ...;

export const Ethereum_Env = instance;
```

**Comment by Niraj-Kamdar** [comment]
## 1. Why don't we inherit properties from env of imported polywrapper? 
### Example:
#### Ethereum polywrapper schema
```graphql
type QueryEnv {
   eth_prop1: String! 
   eth_prop2: Int! @public
}
``` 
#### Ens polywrapper schema
```graphql
type QueryEnv {
   ens_prop1: String! @public
   ens_prop2: Int!
}
``` 
#### User Polywrapper schema
```graphql
#import { QueryEnv } into Ethereum from "..."
#import { QueryEnv } into Ens from ".."

type QueryEnv implements Ethereum_QueryEnv, Ens_QueryEnv {
   prop1: String! 
}
```
#### shape of compiled QueryEnv schema
```graphql
type QueryEnv {
   eth_prop2: Int!
   ens_prop1: String! 
   prop1: String! 
}
```

#### usage 
```ts
import {env} from "./w3";
...
env.eth_prop2
...
env.prop1
```
Usage will be same regardless of module and there won't be any extra step during template generation for propagating env of main polywrapper to imported polywrappers. 

## 2. Why don't we use inherit syntax for `Env` to make inheritance explicit?
### Current Schema Example:
```graphql
#import {Env} from "/some/local/dir";

type QueryEnv {
  prop1: String!
}
```
This implicitly inherits from Env specified in local dir. But having this explicit would be better 
```graphql
#import {Env} from "/some/local/dir";

type QueryEnv implements Env {
  prop1: String! 
}
```

This would also allow us to divide big environment in different files depending on some idea/category/context
```graphql
#import {Env} into Db from "/some/local/dir/db";
#import {Env} into Api from "/some/local/dir/api";

type QueryEnv implements DB_Env, ApiEnv {
  prop1: String! 
}
```

**Comment by dOrgJelli** [comment]
Chatted with @Niraj-Kamdar and @nerfZael today. Our findings are that:
- we will finish the existing PR with the initial spec. Loose schema / environment coupling, all env sanitization will be done at run-time upon invocation.
- In the future, we may apply some of Niraj's suggestions to aggregate dependency environments, to help with the potential run-time errors that occur from the loose coupling.
- We've decided that there is no need to implement the public directive at this point, and can add that in the future as well.
- We've decided to not implement separate environment types for the Query and Mutation modules. This will simplify the implementation, while still allowing it to be added later if need be.
- We discussed moving away from the model of having 1 schema for 2 modules (query, mutation) in the build package, and instead moving to an architecture where it's 1:1 schema for module.
